Add functionality for fallbacks.

seanlip opened the pull request.

seanlip commented:
@jacobdavis11 -- would you mind taking a look at the functionality here, and letting me know what you think? Thanks!

seanlip commented:
@jacobdavis11 -- thank you for the comments! I believe I've addressed everything, except point 6, which is due to a bug in jQueryUI that occurs when the top of the screen is above the top of the viewport -- I've made notes in the code to that effect.

Would you mind taking another look, please? Also /cc @amitdeutsch.

amitdeutsch commented:
LGTM on the UI

jacobdavis11 commented on core/templates/dev/head/components/fallback_editor.html:
```diff
@@ -0,0 +1,66 @@
+               ng-if="isEditable" ng-click="openTriggerEditor()">
+          </div>
+
+          Once the learner has made <[trigger.customization_args.num_submits.value]>
```
It seems odd to hard-code this in here when we're going to have other fallbacks in future.

seanlip replied:
Yeah, I thought about writing a framework for this, and then decided that, at this stage, I'd prefer not to assume anything about other fallbacks in the future yet. I feel that the various fallbacks we've considered are sufficiently different that it's still unclear to me that they all should fit within the same framework, so I'd like to postpone baking that in until we get a better sense for how things work out in practice. The impending editor redesign may shed more light on that, too.

jacobdavis11 commented on core/templates/dev/head/editor/ExplorationEditorAdvancedFeaturesService.js:
```diff
@@ -19,17 +19,24 @@
   var _settings = {
-    areParametersEnabled: false,
-    areGadgetsEnabled: false
+    areFallbacksEnabled: false,
```
You might want to also put these in alphabetical order in the html if there's no reason not to.

seanlip replied:
Done.

jacobdavis11 commented on core/templates/dev/head/editor/StateFallbacks.js:
```diff
@@ -0,0 +1,195 @@
+          '$scope', '$modalInstance', 'editorContextService',
+          function($scope, $modalInstance, editorContextService) {
+
+        $scope.INT_FORM_SCHEMA = {
```
This seems like the kind of thing to handle using the existing rule architecture rather than hard-coding it.

seanlip replied:
The existing rule architecture is not great, actually -- it's rather cobbled together, and I don't want to tie more things to it at present. I'd like to give it an overhaul at some stage.

jacobdavis11 commented on core/tests/protractor/fallbacks.js:
```diff
@@ -0,0 +1,90 @@
+      editor.saveChanges();
+      workflow.publishExploration();
+
+      // Login as another user and verify that the exploration editor does not
```
This doesn't seem like it has anything particularly to do with fallbacks. Perhaps it should go in privileges.js.

seanlip replied:
The checks it makes are to fallbacks-related stuff. Note that there is another test of this kind, and that's currently in editorAndPlayer.js.

I feel that, since fallbacks are an optional component, it's reasonable for everything fallbacks-related to go in the same Protractor test suite.

jacobdavis11 replied:
I think then that all things of this kind should be moved into privileges.js, since "checking the editor works" and "checking the editor cannot be used by unauthorised people" are fundamentally different things. However it seems reasonable to leave this here for now until the tests are refactored.

seanlip replied:
Ack, I agree.

jacobdavis11 commented on core/tests/protractor/fallbacks.js:
```diff
@@ -0,0 +1,90 @@
+      player.expectLatestFeedbackToMatch(forms.toRichText('try again'));
+      player.expectExplorationToNotBeOver();
+      player.submitAnswer('NumericInput', 18);
+      player.expectLatestFeedbackToMatch(forms.toRichText('fallback triggered'));
```
I would add a check for rules that go somewhere taking precedence over a fallback that was about to trigger.

seanlip replied:
Good idea, done.

jacobdavis11 commented on extensions/interactions/Continue/validator.js:
```diff
@@ -26,18 +26,18 @@ oppia.filter('oppiaInteractiveContinueValidator', [
       warningsList.push({
         type: WARNING_TYPES.CRITICAL,
-        message: 'the button text should not be empty.'
+        message: 'The button text should not be empty.'
```
Is this relevant to the current commit?

seanlip replied:
Not particularly. It was something that was irritating me (the warnings showing up in the graph kept starting with a lowercase letter) and I figured (perhaps wrongly) that I might as well fix it since doing so wasn't hard. In retrospect it would probably have been better to leave it to a separate PR.

jacobdavis11 commented on core/tests/protractor_utils/editor.js:
```diff
@@ -673,6 +673,129 @@ var expectCannotAddResponse = function() {
+    }
+  });
+
+  var fallbackElt = element(by.css('.protractor-test-add-fallback-details'));
```
Please be consistent in abbreviations for 'Element'.

seanlip replied:
Oops, thanks -- done.

jacobdavis11 commented on core/templates/dev/head/components/fallbackEditor.js:
```diff
@@ -0,0 +1,76 @@
+          }]
+        };
+
+        $scope.triggerMemento = null;
```
Would it be possible to set up a standard memento service that just gets invoked here, rather than re-implementing it every time mementos are needed?

seanlip replied:
Mm, not easily. We do have a memento for the whole fallbacks object, but that's not granular enough. I think what I've done here is the simplest for now -- but @BenHenning, we should probably revisit how these things are handled when we do the frontend refactor.

jacobdavis11 commented on core/domain/exp_services_test.py:
```diff
@@ -1102,6 +1102,43 @@ def test_update_interaction_answer_groups(self):
+            'Incorrect', '<b>Wrong answer</b>'])
+        self.assertEqual(fallbacks[0].outcome.dest, self.init_state_name)
+
     def test_update_state_invalid_state(self):
```
I would add to this test the feedback destinations cannot be non-existent.

seanlip replied:
Good idea; I added a test.

jacobdavis11 commented:
5. There is still a problem in that if you delete a state the frontend updates correctly, but if you try to save the backend throws an error: "Warning! The fallback destination end is not a valid state.". Xinyu and I speculate that this is because the backend is failing to use the delete command in the changelist to update its versions of the destinations. We suggest that instead the frontend is made responsible for adding to the changelist all the relevant changes to destinations.

7. If there is a path from to END using fallbacks but not one without them the frontend does not show any warnings, but when you attempt to publish the exploration the backend throws an error. "Warning! Please fix the following issues before saving this exploration: 1. The following states are not reachable from the initial state: end2 2. It is impossible to complete the exploration from the following states: First Card".

seanlip commented:
Thank you, @jacobdavis11! I've pushed a commit addressing the comments, PTAL. It also fixes points 5 and 7.

Point 5 happened because I had neglected to update the get_all_outcomes() function in the backend to also consider fallbacks; this was a straightforward change. I would prefer, at this point, not to make larger changes to how the changelist is handled, but this is one thing that @BenHenning and I have been discussing in the plans for the frontend refactor. Not sure what the outcome will be yet, but it will likely increase clarity.

jacobdavis11 commented:
All issues are fixed but I'm afraid I just found another.

_8. Create an exploration with a fallback that activates after two wrong answers. Go to the preview tab, enter a wrong answer once and return immediately to the editor tab. Go again to the preview tab and enter a wrong answer once. The fallback will trigger. If you then go to the main tab and back to the preview tab again and keep submitting answers you will never trigger the fallback. It seems that the wrong answer counter is not being refreshed when the preview tab is entered.

jacobdavis11 commented:
By the way, that worries me a little. It looks like parameters are correctly refreshed but I don't know if that's due to the way the player is embedded into the preview tab, or to a special case invocation. Ideally it would be the former, which would also handle things like wrong answer counts.

seanlip commented:
Thanks for catching that -- this is now fixed. I added an [additional reset](https://github.com/oppia/oppia/commit/4f0780af0d2b3c90689077c71f2968f892897abd) to init() -- previously this was done only when the state changed, which works fine in the learner view but fails in editor preview mode.

jacobdavis11 commented:
Looks good.
